### ISSUE TITLE
Implement mail system

### ISSUE THREAD

**Issue by bhavanki** [bhavanki/grounds#5]
Implement a softcode system for mail between players.

**Comment by raelik** [comment]
In order to be able to handle the background tasks necessary for a mail system, you'd need to expand the functionality of the Extension class.  Right now, you can only create scripted commands and listeners in an extension, and use attributes to manage the state of the extension.  What you would need is a specific attribute to store an initialization script that gets run just before the extension is registered in LoadCommand.  In addition to performing any initialization that might be necessary when a universe is loaded (say to transfer any mail that might have been waiting in the queue while the universe was unloaded), it could also start any background threads that might be needed.  This would necessitate adding an instance variable to Extension to hold a reference to a thread or ExecutorService (an ExecutorService would probably be static instead), and possibly a Map to store references to any shared state that may be desirable to expose. Anything of that nature should probably be an Atomic class or something like a Queue, or at the very least synchronized when modified.

**Comment by bhavanki** [comment]
@raelik Thanks for your thoughts! (Yes, it took me this long to get back to working on this project ...)

I tried writing a mail system in softcode in July / August of 2020 (!) and quickly found that it was going to be pretty complicated, and set it aside. When I decided to hop back in and work some more here, I decided to just go an easier and more fun way and hardcode the system instead. What I wrote probably could serve as a template for a softcode re-implementation ... but I'm not excited to try it and I don't know if it will ever matter. (Meanwhile, the missing mail system was probably the biggest hole in this whole project!)

On the subject of threading: One of the design principles I started with is to store all of the game state in the universe, and serialize all changes to the state. That's why there's just one command executor, and commands are the only way to modify state. (Way under the hood, scripts are forced to use commands, too.) With this, you never have to worry about simultaneous state changes when, say, ten players are all doing stuff. Mail data is part of the game state - because it's player-to-player - so mail commands for the data are enough, without needing extra executors or other synchronization mechanisms.

It's maybe a little weird, then, for the game to treat sending a mail message like performing an action in-game. My hope is that the actions are simple enough that there is no perceptible performance issue. If one arises, there are some ways, like what you suggest, to add some more concurrency or prioritization.
